### ISSUE TITLE
ZeroDataStateNotifications component should not attempt data requests for recoverable modules in view-only context

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#5505]
## Bug Description

The `ZeroDataStateNotifications` component calls the `isGatheringData` and `hasZeroData` selectors for the Analytics and Search Console modules, and each of these selectors in turn relies on a request to retrieve report data for the corresponding module. However when in view-only context and the modules are in a recoverable state, the user does not have permission to request module data, so these requests fail and a JS error is seen in the console.

We should avoid calling these selectors when in view-only context for recoverable modules.

## Steps to reproduce

1. Share Analytics and/or Search Console and put into a recoverable state.
2. Navigate to the view-only Dashboard as a user the modules are shared with.
3. Note that failed requests are made to the module report endpoints and JS errors are seen in the console.

## Screenshots

![image.png](https://images.zenhubusercontent.com/6133937d41f397090b8344a7/19392e72-b3d7-4ba2-90cc-5eaa79b8e6ab)

## Additional Context

 - PHP Version: any
 - OS: any
 - Browser: any
 - Plugin Version: 1.77.0
 - Device: any

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- When in view-only context and either Analytics or Search Console is in a recoverable state:
  - Viewing the `ZeroDataStateNotifications` component (by navigating to the Main or Entity Dashboard) should not result in any requests for the recoverable module's report data being made.
  - As a result no related errors should appear in the JS console.

## Implementation Brief

- In `assets/js/components/notifications/ZeroDataStateNotifications/index.js`:
    - Use the `useSelect` hook to call the `getRecoverableModules` selector from the `core/modules` store and assign it to a new variable named `recoverableModules`.
    - Check if the keys of the object `recoverableModules` includes `analytics` and assign the boolean value to a new variable named `isAnalyticsRecoverable`.
    - Check if the keys of the object `recoverableModules` includes `search-console` and assign the boolean value to a new variable named `isSearchConsoleRecoverable`.
    - Locate the `isGatheringData` and `hasZeroData` selector calls from the `modules/analytics` and `modules/search-console` stores. They are currently assigned to the following variables:
        - `analyticsGatheringData`
        - `searchConsoleGatheringData`
        - `analyticsHasZeroData`
        - `searchConsoleHasZeroData`
    - Use the `useViewOnly` hook (currently assigned to the `viewOnly` variable) to check if it is a view-only context.
    - For `analyticsGatheringData` and `analyticsHasZeroData`, only call the selectors if beside the current conditions, the following conditions are met:
        - `viewOnly` is falsey. **OR**
        - `viewOnly` is truthy but `isAnalyticsRecoverable` is falsey.
    - For `searchConsoleGatheringData` and `searchConsoleHasZeroData`, only call the selectors if beside the current conditions, the following conditions are met:
        - `viewOnly` is falsey. **OR**
        - `viewOnly` is truthy but `isSearchConsoleRecoverable` is falsey.

### Test Coverage

No tests need to be added/updated.

## QA Brief

- Ensure that the `dashboardSharing` & `zeroDataStates` feature flags are enabled.
- Sign in as an admin and connect Site Kit with Analytics and share Analytics + Search Console via the Dashboard Sharing settings dialog.
- Sign in as a 2nd admin and connect Site Kit.
- Sign in as the 1st admin again and disconnect Site Kit.
- Sign in as another user and view the view-only dashboard.
- Clear local and session storage.
- Open the `Browser Developer Tools -> Network` tab.
- Ensure that no requests are made to the following endpoints:
    - `/wp-json/google-site-kit/v1/modules/search-console/data/searchanalytics`
    - `/wp-json/google-site-kit/v1/modules/analytics/data/report`
- Check the browser console and ensure that no errors similar to the ones shown in the issue description screenshots show up.

## Changelog entry

* Update the `ZeroDataStateNotifications` component to not attempt data requests for recoverable modules in view-only context.

**Comment by aaemnnosttv** [comment]
@techanvil this is probably what you meant, but right now it reads a bit like the change is coupled to both modules in a recoverable state when it really applies to both independently. Let's reword it to be more clear about this, then it should be good to go 👍

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, that is a good point and I've tweaked the AC accordingly.

**Comment by techanvil** [comment]
Hi @nfmohit, this IB is technically fine and I am happy to :white_check_mark: it as it is.

However, I would say that it's a bit on the verbose side and pretty much transliterates code into text, which is a bit beyond what an IB needs to achieve. There are some details we could quite happily do without, and, for the sake of illustration, I'd suggest that it could be reduced to something like the following:

- In `assets/js/components/notifications/ZeroDataStateNotifications/index.js`:
    - Retrieve the set of recoverable modules from the `getRecoverableModules` selector. This returns an object keyed by module slug - if a module is recoverable its slug will be present.
    - Locate the `isGatheringData` and `hasZeroData` selector calls from the `modules/analytics` and `modules/search-console` stores.
    - For each of these selector calls, short-circuit the corresponding `useInViewSelect` to return `false` when `viewOnly` is `true` **AND** the corresponding module is recoverable.

**Comment by techanvil** [comment]
IB :white_check_mark:

**Comment by nfmohit** [comment]
> However, I would say that it's a bit on the verbose side and pretty much transliterates code into text, which is a bit beyond what an IB needs to achieve. There are some details we could quite happily do without, and, for the sake of illustration, I'd suggest that it could be reduced to something like the following:

Thank you very much for the kind and detailed feedback here, @techanvil. Looking at your example, I realise that the IB I added could indeed be much more concise. I'll keep it in mind for the future ones. Thank you!

**Comment by techanvil** [comment]
Cheers Nahid!

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- No requests are made to the following endpoints:
  - /wp-json/google-site-kit/v1/modules/search-console/data/searchanalytics
  - /wp-json/google-site-kit/v1/modules/analytics/data/report
- No errors similar to the ones shown in the issue description screenshots show up in the browser console. 

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/187681805-55934c88-1447-42a1-92eb-e90ba8eeed15.png)
![image](https://user-images.githubusercontent.com/73545194/187681828-ead69972-3f12-4f87-ac07-3ec9322275b6.png)
![image](https://user-images.githubusercontent.com/73545194/187681853-093a4a39-3d1b-4207-9a93-bbb48e16f300.png)

</details>
